### PR TITLE
refactor: Adjust credential endpoints permissions

### PR DESCRIPTION
**Pull request by krynble**

### CODE REVIEW

File: packages/cli/src/UserManagement/PermissionChecker.ts [hunk 1]
```diff
@@ -26,12 +26,19 @@ export class PermissionChecker {
 		// allow if all creds used in this workflow are a subset of
 		// all creds accessible to users who have access to this workflow
 
-		const workflowSharings = await Db.collections.SharedWorkflow.find({
-			relations: ['workflow'],
-			where: { workflow: { id: Number(workflow.id) } },
-		});
-
-		const workflowUserIds = workflowSharings.map((s) => s.userId);
+		let workflowUserIds = [] as string[];
```
Comment (ivov) [inline packages/cli/src/UserManagement/PermissionChecker.ts:29]:
```suggestion
		let workflowUserIds: string[] = [];
```

File: packages/cli/src/UserManagement/PermissionChecker.ts [hunk 2]
```diff
@@ -26,12 +26,19 @@ export class PermissionChecker {
...
+
+		// unsaved workflows have no id, so only get credentials for current user
```
Comment (ivov) [inline packages/cli/src/UserManagement/PermissionChecker.ts:31]: Nit: The comment refers to the else branch, so I'd place that case first. Also generally I find it more readable to have shorter consequents first.

File: packages/cli/src/credentials/credentials.service.ee.ts [hunk 3]
```diff
@@ -25,6 +25,36 @@ export class EECredentialsService extends CredentialsService {
...
+	/**
+	 * Retrieve the sharing that matches a user and a credential.
+	 */
+	static async getSharing(
+		user: User,
+		credentialId: number | string,
+		relations: string[] = ['credentials'],
+		{ allowGlobalOwner } = { allowGlobalOwner: true },
+	): Promise<SharedCredentials | undefined> {
+		const options: FindOneOptions = {
```
Comment (ivov) [inline packages/cli/src/credentials/credentials.service.ee.ts:37]: Prevent `@typescript-eslint/no-unsafe-member-access` below.
```suggestion
		const options: FindOneOptions<SharedCredentials> & { where: ObjectLiteral } = {
```

File: packages/cli/src/credentials/credentials.service.ee.ts [hunk 4]
```diff
@@ -25,6 +25,36 @@ export class EECredentialsService extends CredentialsService {
 		return { ownsCredential: true, credential };
 	}
 
+	/**
+	 * Retrieve the sharing that matches a user and a credential.
+	 */
+	static async getSharing(
+		user: User,
+		credentialId: number | string,
+		relations: string[] = ['credentials'],
```
Comment (ivov) [inline packages/cli/src/credentials/credentials.service.ee.ts:34]: Nit: No relations requires an empty array, not ideal, perhaps consolidated as an option with the next param. (Comes from before, feel free to disregard.)

File: packages/cli/src/workflows/workflows.controller.ee.ts [hunk 5]
```diff
@@ -247,13 +247,14 @@ EEWorkflowController.post(
...
-			workflow,
-			workflow.id.toString(),
-			req.user,
-		);
-
-		req.body.workflowData.nodes = safeWorkflow.nodes;
+		if (workflow.id !== undefined) {
+			const safeWorkflow = await EEWorkflows.preventTampering(
+				workflow,
+				workflow.id?.toString(),
```
Comment (ivov) [inline packages/cli/src/workflows/workflows.controller.ee.ts:253]:
```suggestion
				workflow.id.toString(),
```

File: packages/cli/src/credentials/credentials.service.ts [hunk 6]
```diff
@@ -52,11 +55,21 @@ export class CredentialsService {
 
 		// if member, return credentials owned by or shared with member
 
-		const userSharings = await Db.collections.SharedCredentials.find({
+		const whereConditions: FindManyOptions = {
```
Comment (ivov) [inline packages/cli/src/credentials/credentials.service.ts:58]: Something is very odd here.

I expect this should be `FindManyOptions<SharedCredentials> & { where: ObjectLiteral }` and no assertion below, much like the other fix, but for some reason `where` here is not being narrowed.

This is not worth more than a few minutes, so I set it aside for now. Feel free to disregard. Also below.

File: packages/cli/src/credentials/credentials.service.ts [hunk 7]
```diff
@@ -52,11 +55,21 @@ export class CredentialsService {
...
 				user,
 			},
-		});
+		};
+
+		if (options?.roles?.length) {
+			whereConditions.where = {
+				...whereConditions.where,
+				role: { name: In(options.roles) },
+			} as FindOneOptions;
```
Comment (ivov) [inline packages/cli/src/credentials/credentials.service.ts:68]: This is `FindOneOptions` but the query is for many.

**Review by ivov — COMMENTED** [review]
Confirmed working, all tests passing on all three DBs.

**Review by ivov — APPROVED** [review]

**Comment by janober** [comment]
Got released with `n8n@0.204.0`

**Pull request merged**